src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+                PM.add(createVerifierPass());
+#endif
+            }
+            if (TM.addPassesToEmitMC(PM, Ctx, ObjStream))
+                llvm_unreachable("Target does not support MC emission.");
+
+            CompileLayer = std::unique_ptr<CompileLayerT>{new CompileLayerT(ObjectLayer,
+                [&](Module &M) {
+                    PM.run(M);
+                    std::unique_ptr<MemoryBuffer> ObjBuffer(
+                        new ObjectMemoryBuffer(std::move(ObjBufferSV)));
+                    ErrorOr<std::unique_ptr<object::ObjectFile>> Obj =
+                        object::ObjectFile::createObjectFile(ObjBuffer->getMemBufferRef());
+
+                    // TODO: Actually report errors helpfully.
+                    if (Obj)
+                        return OwningObj(std::move(*Obj), std::move(ObjBuffer));
+                    return OwningObj(nullptr, nullptr);
+                }
+            )};
+            // Make sure SectionMemoryManager::getSymbolAddressInProcess can resolve
+            // symbols in the program as well. The nullptr argument to the function
+            // tells DynamicLibrary to load the program, not a library.
+
+            std::string *ErrorStr = nullptr;
+            if (sys::DynamicLibrary::LoadLibraryPermanently(nullptr, ErrorStr))
+                report_fatal_error("FATAL: unable to dlopen self\n" + *ErrorStr);
+        }
+
+    std::string mangle(const std::string &Name) {
+        std::string MangledName;
+        {
+            raw_string_ostream MangledNameStream(MangledName);
+            Mangler::getNameWithPrefix(MangledNameStream, Name, DL);
+        }
+        return MangledName;
+    }
+
+    void addGlobalMapping(StringRef Name, void *Addr) {
+       GlobalSymbolTable[mangle(Name)] = Addr;
+    }
+
+    ModuleHandleT addModule(Module *M) {
+        // We need a memory manager to allocate memory and resolve symbols for this
+        // new module. Create one that resolves symbols by looking back into the
+        // JIT.
+        auto Resolver = orc::createLambdaResolver(
+                          [&](const std::string &Name) {
+                            // TODO: consider moving the FunctionMover resolver here
+                            // Step 0: ObjectLinkingLayer has checked whether it is in the current module
+                            // Step 1: Check against list of known external globals
+                            GlobalSymbolTableT::const_iterator pos = GlobalSymbolTable.find(Name);
+                            if (pos != GlobalSymbolTable.end())
+                                return RuntimeDyld::SymbolInfo((intptr_t)pos->second, JITSymbolFlags::Exported);
+                            // Step 2: Search all previously emitted symbols
+                            if (auto Sym = findSymbol(Name))
+                              return RuntimeDyld::SymbolInfo(Sym.getAddress(),
+                                                             Sym.getFlags());
```

Comment:
in the FunctionMover, I thought you indicated this search may be expensive and `O(n)` in the number of modules that have been emitted. If so, and if you're still looking for places where performance is potentially being lost, this may be one place to consider.